tonic-types/src/lib.rs

```diff
@@ -13,8 +23,391 @@
+    /// [`WithErrorDetails::get_error_details_vec`] or
+    /// [`WithErrorDetails::get_error_details`].
+    /// # Examples
+    ///
+    /// ```
+    /// use tonic::{Status, Response};
+    /// use tonic_types::{ErrorDetail, WithErrorDetails};
+    ///
+    /// fn handle_request_result<T>(req_result: Result<Response<T>, Status>) {
+    ///     match req_result {
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             match status.check_error_details_vec() {
+    ///                 Ok(err_details) => {
+    ///                     // Handle extracted details
+    ///                 }
+    ///                 Err(decode_error) => {
+    ///                     // Handle decode_error
+    ///                 }
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn check_error_details_vec(&self) -> Result<Vec<ErrorDetail>, DecodeError>;
+
+    /// Get a vector of [`ErrorDetail`] enums from `tonic::Status`. If some
+    /// `prost::DecodeError` occurs, an empty vector will be returned.
+    /// # Examples
+    ///
+    /// ```
+    /// use tonic::{Status, Response};
+    /// use tonic_types::{ErrorDetail, WithErrorDetails};
+    ///
+    /// fn handle_request_result<T>(req_result: Result<Response<T>, Status>) {
+    ///     match req_result {
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             let err_details = status.get_error_details_vec();
+    ///             for err_detail in err_details.iter() {
+    ///                  match err_detail {
+    ///                     ErrorDetail::BadRequest(bad_request) => {
+    ///                         // Handle bad_request details
+    ///                     }
+    ///                     // ...
+    ///                     _ => {}
+    ///                  }
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn get_error_details_vec(&self) -> Vec<ErrorDetail>;
+
+    /// Get first [`BadRequest`] details found on `tonic::Status`, if any. If
+    /// some `prost::DecodeError` occurs, returns `None`.
+    /// # Examples
+    ///
+    /// ```
+    /// use tonic::{Status, Response};
+    /// use tonic_types::{WithErrorDetails};
+    ///
+    /// fn handle_request_result<T>(req_result: Result<Response<T>, Status>) {
+    ///     match req_result {
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             if let Some(bad_request) = status.get_details_bad_request() {
+    ///                 // Handle bad_request details
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn get_details_bad_request(&self) -> Option<BadRequest>;
+}
+
+impl WithErrorDetails for tonic::Status {
+    fn with_error_details(code: Code, message: impl Into<String>, details: ErrorDetails) -> Self {
+        let message: String = message.into();
+
+        let mut conv_details: Vec<Any> = Vec::with_capacity(10);
```

Comment:
Where did you get the capacity of 10?

Reply:
It is the total number of standard error message types, the final `ErrorDetails` struct will have 10 fields